File: dawn_python/surf.py
from collections import defaultdict

def build(keys):
    root = defaultdict(lambda: defaultdict())
    for key in keys:
        curr = root
        for char in key:
            curr = curr.setdefault(char, defaultdict())
    return root

def lookup(trie, key):
    curr = trie
    for char in key:
        if char not in curr:
            return False
        curr = curr[char]
    return True

File: dawn_rust/src/surf.rs
use std::collections::HashMap;

type Node = HashMap<char, Node>;

fn build(keys: &[&str]) -> Node {
    let mut root = Node::new();
    for key in keys {
        let mut curr = &mut root;
        for c in key.chars() {
            curr = curr.entry(c).or_insert(Node::new());
        }
    }
    root
}

fn lookup(trie: &Node, key: &str) -> bool {
    let mut curr = trie;
    for c in key.chars() {
        match curr.get(&c) {
            Some(next) => curr = next,
            None => return false,
        }
    }
    true
}